ark/jsonschema/object.ts

```diff
@@ -0,0 +1,336 @@
+					})
+		})
+	}
+	return predicates
+}
+
+const parsePatternProperties = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	if (!("patternProperties" in jsonSchema)) return
+
+	const patternProperties = Object.entries(jsonSchema.patternProperties).map(
+		([key, value]) =>
+			[new RegExp(key), innerParseJsonSchema.assert(value)] as const
+	)
+
+	// Ensure that the schema for any property is compatible with any corresponding patternProperties
+	patternProperties.forEach(([pattern, parsedPatternPropertySchema]) => {
+		Object.entries(jsonSchema.properties ?? {}).forEach(
+			([property, schemaForProperty]) => {
+				if (!pattern.test(property)) return
+
+				const parsedPropertySchema =
+					innerParseJsonSchema.assert(schemaForProperty)
+
+				if (!parsedPropertySchema.overlaps(parsedPatternPropertySchema)) {
+					ctx.reject({
+						message: `property ${property} must have a schema that overlaps with the patternProperty ${pattern}`
+					})
+				}
+			}
+		)
+	})
+
+	// NB: We don't validate compatability of schemas for overlapping patternProperties
+	// since getting the intersection of regexes is inherenetly difficult.
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.entries(data).forEach(([dataKey, dataValue]) => {
+			patternProperties.forEach(([pattern, parsedJsonSchema]) => {
+				if (pattern.test(dataKey) && !parsedJsonSchema.allows(dataValue)) {
+					errors.push(
+						ctx.reject({
+							actual: dataValue,
+							expected: `${parsedJsonSchema.description} as property ${dataKey} matches patternProperty ${pattern}`
+						})
+					)
+				}
+			})
+		})
+		return errors.length === 0
+	}
+}
+
+const parsePropertyNames = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	if (!("propertyNames" in jsonSchema)) return
+
+	const propertyNamesValidator = innerParseJsonSchema.assert(
+		jsonSchema.propertyNames
+	)
+
+	if (
+		"domain" in propertyNamesValidator.json &&
+		propertyNamesValidator.json.domain !== "string"
+	) {
+		ctx.reject({
+			path: ["propertyNames"],
+			actual: `a schema for validating a ${propertyNamesValidator.json.domain as string}`,
+			expected: "a schema for validating a string"
+		})
+	}
+
+	return (data: object, ctx: TraversalContext) => {
+		const errors: false[] = []
+
+		Object.keys(data).forEach(key => {
+			if (!propertyNamesValidator.allows(key)) {
+				errors.push(
+					ctx.reject({
+						message: `property ${key} doesn't adhere to the propertyNames schema of ${propertyNamesValidator.description}`
+					})
+				)
+			}
+		})
+		return errors.length === 0
+	}
+}
+
+const parseRequiredAndOptionalKeys = (
+	jsonSchema: JsonSchema.ObjectSchema,
+	ctx: TraversalContext
+) => {
+	const optionalKeys: string[] = []
+	const requiredKeys: string[] = []
+	if ("properties" in jsonSchema) {
+		if ("required" in jsonSchema) {
+			if (jsonSchema.required.length !== new Set(jsonSchema.required).size) {
+				ctx.reject({
```

Comment:
I'd write this as a standalone function called `duplicateElementsOf` that iterates over an array and each time it encounters an element it's seen, it adds a `DuplicateData` like `{ value: unknown, indices: number[] }`.

This could be added to `@ark/util` in `arrays.ts` since it would be general, but it would also help to write a much more descriptive error message here

Reply:
Done across 9826c00b and 5272e44c